### PR TITLE
[WFLY-14347] Allow to configure module for custom validation classes

### PR DESCRIPTION
**Pull request by tadamski**
https://issues.redhat.com/browse/WFLY-14347

This issue is an implementation of:
https://issues.redhat.com/browse/EAP7-1189

### CODE REVIEW

File: connector/src/main/java/org/jboss/as/connector/subsystems/datasources/Constants.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  * JBoss, Home of Professional Open Source.
- * Copyright 2010, Red Hat, Inc., and individual contributors
+ * Copyright 2021, Red Hat, Inc., and individual contributors
```
Comment (darranl) [inline connector/src/main/java/org/jboss/as/connector/subsystems/datasources/Constants.java:3]: We don't normally go for copyright updates for minor updates to files.

**Review by darranl — DISMISSED** [review]
Is this missing schema updates and transformers etc...?

Have the relevant models already been bumped for WildFly 25 or is that also needed?

Do we have an RFE to track the changes?

File: connector/src/main/java/org/jboss/as/connector/util/ModelNodeUtil.java [hunk 2]
```diff
@@ -41,16 +65,36 @@ public static String getResolvedStringIfSetOrGetDefault(final OperationContext c
...
+                moduleName = getResolvedStringIfSetOrGetDefault(operationContext, dataSourceNode, moduleNameAttribute);
+            }
+
+            ClassLoader moduleClassLoader = null;
+            if (moduleName != null) {
+                try {
+                    final ModuleIdentifier exceptionSorterModuleId = ModuleIdentifier.create(moduleName);
+                    moduleClassLoader = Module.getCallerModuleLoader().loadModule(exceptionSorterModuleId).getClassLoader();
+                } catch (ModuleLoadException exception) {
+                    throw new OperationFailedException("Wrong module name", exception);
```
Comment (bstansberry) [inline connector/src/main/java/org/jboss/as/connector/util/ModelNodeUtil.java:90]: @tadamski This needs to be an i18n message, and should include the faulty module name.

File: connector/src/main/java/org/jboss/as/connector/util/ModelNodeUtil.java [hunk 3]
```diff
@@ -41,16 +65,36 @@ public static String getResolvedStringIfSetOrGetDefault(final OperationContext c
...
+
+            String moduleName = null;
+            if (moduleNameAttribute != null) {
+                moduleName = getResolvedStringIfSetOrGetDefault(operationContext, dataSourceNode, moduleNameAttribute);
+            }
+
+            ClassLoader moduleClassLoader = null;
+            if (moduleName != null) {
+                try {
+                    final ModuleIdentifier exceptionSorterModuleId = ModuleIdentifier.create(moduleName);
```
Comment (bstansberry) [inline connector/src/main/java/org/jboss/as/connector/util/ModelNodeUtil.java:87]: ModuleIdentifier is deprecated. You should just use moduleName.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+import static org.jboss.as.controller.client.helpers.ClientConstants.ADD;
+import static org.jboss.as.controller.client.helpers.ClientConstants.OP;
+import static org.jboss.as.controller.client.helpers.ClientConstants.OPERATION_HEADERS;
+import static org.jboss.as.controller.client.helpers.ClientConstants.OP_ADDR;
+import static org.jboss.as.controller.client.helpers.ClientConstants.OUTCOME;
+import static org.jboss.as.controller.client.helpers.ClientConstants.REMOVE_OPERATION;
+import static org.jboss.as.controller.client.helpers.ClientConstants.SUBSYSTEM;
+import static org.jboss.as.controller.client.helpers.ClientConstants.SUCCESS;
+import static org.jboss.as.test.shared.integration.ejb.security.PermissionUtils.createPermissionsXmlAsset;
```
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java:74]: The ordering if import sections here is backwards.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        try {
+            if (!connection.isClosed()) {
+                connection.close();
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public static class Setup implements ServerSetupTask {
```
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java:245]: Please use org.jboss.as.test.module.util.TestModule for all this

**Review by bstansberry — CHANGES_REQUESTED** [review]
@tadamski Some minor points.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java (hunk 4 above)
Comment (bstansberry) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java:74]: Correcting this is a nice to have; can come later.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java (hunk 5 above)
Comment (bstansberry) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java:245]: Correcting this is a nice to have; can come later.

File: connector/src/main/java/org/jboss/as/connector/util/ModelNodeUtil.java (hunk 2 above)
Comment (tadamski) [reply connector/src/main/java/org/jboss/as/connector/util/ModelNodeUtil.java]: fixed

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java (hunk 4 above)
Comment (tadamski) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java]: fixed

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java (hunk 5 above)
Comment (tadamski) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jca/classloading/custommodule/DatasourceCustomModulesTestCase.java:241]: fixed

**Review by bstansberry — COMMENTED** [review]
This all LGTM except the subsystem test failure CI is reporting, which can be corrected with a simple update like https://github.com/tadamski/wildfly/pull/2

**Review by bstansberry — APPROVED** [review]

**Pull request merged**